Enable window docking

bobtista opened the pull request.

bobtista committed:
add keyword for build

bobtista committed:
enable docking to the main window

bobtista committed:
disable multiviewport, so main window is the viewport, other windows stay on top

bobtista committed:
explicitly set enum values for AsmMatchValueEx to address build error

bobtista committed:
fix: add imgui_config.h to properly define math operators before imgui includes

bobtista committed:
refactor: use std::underlying_type for enum instead of explicitly setting integer values

bobtista committed:
fix: lambda signatures and window state tracking in ImGuiApp

- Remove WorkQueueResultPtr parameter from chain() lambdas
- Replace m_imguiHasOpenWindow with m_imguiComparisonWindowOpened

xezon commented on src/imguiclient/imgui_config.h:
This file should no longer exist.

xezon commented on src/imguiclient/imguiapp.cpp:
```diff
@@ -1410,66 +1404,77 @@ std::string ImGuiApp::create_time_string(std::chrono::time_point<std::chrono::sy
-
-    ImScoped::Window window("File Manager", p_open, ImGuiWindowFlags_MenuBar);
+    ImGuiWindowFlags window_flags = ImGuiWindowFlags_MenuBar;
+    ImScoped::Window window("File Manager", p_open, window_flags);
```
Why was this code changed? Was the merge conflict incorrectly resolved?

bobtista replied:
Most likely yes, I tried choosing your branch at each conflict, but I may have missed this one

bobtista replied:
Fixed.

xezon commented on src/imguiclient/imguiapp.cpp:
```diff
@@ -1492,8 +1497,11 @@ void ImGuiApp::ComparisonManagerWindows()
-        if (descriptor.m_imguiComparisonWindowOpened)
+        ImScoped::Window window(title.c_str(), &descriptor.m_imguiComparisonWindowOpened);
+        ImScoped::ID id(i);
+        if (window.IsContentVisible && descriptor.m_imguiComparisonWindowOpened)
```
Why was this code changed? Was the merge conflict incorrectly resolved?

bobtista replied:
Yes, I put it back

xezon commented on src/util.h:
```diff
@@ -90,7 +90,7 @@ constexpr int compare_nocase(std::string_view str1, std::string_view str2)
 // The output string will always fit the formatted string in its entirety.
 template<typename String, typename... Args>
-void assign_format(String &output, fmt::format_string<Args...> format, Args &&... args)
+void assign_format(String &output, fmt::format_string<Args...> format, Args &&...args)
```
The formatting has changed here unnecessarily.

bobtista replied:
My formatter runs on save, so these things are annoying but only once - unless you use a different linter/formatter.

xezon replied:
OmniBlade has told us to use the clang-format 10.0.0.0 which is part of LLVM. You can install that and manually link clang-format 10.0.0.0 in your IDE if it allows to do that (Visual Studio for Windows does).

xezon commented on src/imguiclient/imguiapp.cpp:
```diff
@@ -1410,66 +1404,77 @@ std::string ImGuiApp::create_time_string(std::chrono::time_point<std::chrono::sy
     {
-        ImGui::SetWindowPos("main", m_windowPos);
-        ImGui::SetWindowSize("main", ImVec2(m_windowSize.x, 0.f));
+        ImGuiID dockspace_id = ImGui::GetID("MyDockSpace");
```
Can put a better name for the Dockspace. Maybe "DockSpaceId"

xezon replied:
I meant the string, not the variable name.

bobtista replied:
Done

xezon commented on src/imguiclient/imguiapp.cpp:
```diff
@@ -1410,66 +1404,77 @@ std::string ImGuiApp::create_time_string(std::chrono::time_point<std::chrono::sy
-        ImScoped::MenuBar menu_bar;
-        if (menu_bar.IsOpen)
+    // Menu bar code
+    if (ImGui::BeginMenuBar())
```
I suggest put this block into a new function to reduce complexity in this function.

xezon commented on src/imguiclient/imguiapp.cpp:
```diff
@@ -1410,66 +1404,77 @@ std::string ImGuiApp::create_time_string(std::chrono::time_point<std::chrono::sy
     bool window_open = true;
-    ImScoped::Window window("main", &window_open, window_flags);
-    if (window.IsContentVisible)
+    ImGui::Begin("DockSpace", &window_open, window_flags);
```
Why was `ImScoped::Window window("main", &window_open, window_flags);` replaced with this `ImGui::Begin("DockSpace", &window_open, window_flags);`? It no longer checks the return and the ImScoped looked just fine.

xezon commented on src/imguiclient/imguiapp.cpp:
```diff
@@ -1410,66 +1404,77 @@ std::string ImGuiApp::create_time_string(std::chrono::time_point<std::chrono::sy
+
+    // Set window position and size to cover the entire viewport
+    ImGui::SetNextWindowPos(viewport->WorkPos);
+    ImGui::SetNextWindowSize(viewport->WorkSize);
```
Does this mean we no longer need the `m_windowPos` and `m_windowSize`?

bobtista replied:
Yes, removed them from here and from imguiglfw

xezon replied:
You would also need to remove it from imguiwin32.

bobtista replied:
Done

xezon commented on src/imguiclient/imguiapp.cpp:
```diff
@@ -1410,66 +1404,77 @@ std::string ImGuiApp::create_time_string(std::chrono::time_point<std::chrono::sy
 void ImGuiApp::BackgroundWindow()
 {
-    // #TODO: Make the background dockable somehow. There is a example in ImGui Demo > Examples > Dockspace
+    // Configure flags for the background window
```
Not every line of code needs to be commented if the code is self explanatory.

xezon commented on src/imguiclient/imguiapp.cpp:
```diff
@@ -259,7 +259,7 @@ ImGuiStatus ImGuiApp::init(const CommandLineOptions &clo)
     io.ConfigFlags |= ImGuiConfigFlags_NavEnableGamepad; // Enable Gamepad Controls
     io.ConfigFlags |= ImGuiConfigFlags_DockingEnable; // Enable Docking
-    io.ConfigFlags |= ImGuiConfigFlags_ViewportsEnable; // Enable Multi-Viewport / Platform Windows
+    // io.ConfigFlags |= ImGuiConfigFlags_ViewportsEnable; // Enable Multi-Viewport / Platform Windows
```
What did this do?

bobtista replied:
Windows will stay contained within the main application window. Without this line commented, I kept having windows go behind the main window

xezon replied:
Ah ok nice. I was also wondering about this thing. Good that you found it. Perhaps add a comment what this does.

bobtista replied:
Done

xezon commented on src/imguiclient/imguiapp.cpp:
```diff
@@ -1410,66 +1404,77 @@ std::string ImGuiApp::create_time_string(std::chrono::time_point<std::chrono::sy
+        }
 
-            {
-                ImScoped::Menu menu("Tools");
```
Why was `ImScoped::Menu` replaced by `ImGui::BeginMenu`?

The point of the ImScoped is to auto close the Begin on end of scope, so that `ImGui::EndMenu()` does not need to be written manually.

bobtista committed:
refactor: improve ImGui window management

- Replace manual ImGui::Begin/End with ImScoped::Menu for safer RAII
- Remove manual window position/size tracking in favor of viewport system
- Clean up GLFW window management code

bobtista committed:
remove unnecessary comments

bobtista committed:
refactor: use ImScoped::MenuBar for consistent RAII menu handling

bobtista committed:
fix: remove unused import

xezon commented on src/imguiclient/imguiapp.cpp:
```diff
@@ -1422,23 +1414,40 @@ void ImGuiApp::BackgroundWindow()
                 {
                     if (ImGui::MenuItem("Exit"))
                     {
-                        // Will wait for all work to finish and then shutdown the app.
```
Can keep this comment.

xezon commented on src/imguiclient/imguiapp.cpp:
```diff
@@ -1467,8 +1476,6 @@ void ImGuiApp::BackgroundWindow()
 
 void ImGuiApp::FileManagerWindow(bool *p_open)
 {
-    ImGui::SetNextWindowSizeConstraints(ImVec2(400, 300), ImVec2(FLT_MAX, FLT_MAX));
```
Keep this code.

xezon commented on src/imguiclient/imguiapp.cpp:
```diff
@@ -1492,10 +1499,10 @@ void ImGuiApp::ComparisonManagerWindows()
     for (size_t i = 0; i < count; ++i)
     {
         ProgramComparisonDescriptor &descriptor = *m_programComparisons[i];
+        const std::string title = fmt::format("Assembler Comparison {:d}", descriptor.m_id);
```
This needs to be removed.

xezon commented on src/imguiclient/imguiapp.cpp:
```diff
@@ -1492,10 +1499,10 @@ void ImGuiApp::ComparisonManagerWindows()
 
         if (descriptor.m_imguiComparisonWindowOpened)
         {
-            // Main window
```
Can keep this comment.

bobtista committed:
rename variable and ID

bobtista committed:
address pr comments, restore removed comments, lines of code

bobtista committed:
refactor: remove manual window positioning from Win32 implementation

- Remove GetNativeWindowClientArea helper function
- Remove window position/size setting calls

xezon committed:
Revert formatting changes
